objectbox/test/box_test.dart

```diff
@@ -45,6 +45,94 @@ void main() {
     expect(object.id, equals(putId)); // ID on the object is updated
   });
 
+  test('.putAsync', () async {
+    final box = store.box<TestEntityNonRel>();
+    final a = TestEntityNonRel.filled(id: 0);
+    final b = TestEntityNonRel.filled(id: 0);
+    Future<int> aId = box.putAsync(a);
+    expect(a.id, 0);
+    Future<int> bId = box.putAsync(b);
+    expect(b.id, 0);
+    expect(await aId, 1);
+    expect(await bId, 2);
+    expect(a.id, 1);
+    expect(b.id, 2);
+  });
+
+  test('.putAsync failures', () async {
+    final box = store.box<TestEntity2>();
+    expect(
+        () => box
+            .putAsync(TestEntity2(), mode: PutMode.update)
+            .timeout(defaultTimeout),
+        throwsA(predicate(
+            (ArgumentError e) => e.toString().contains('ID is not set'))));
+
+    expect(
+        await box
+            .putAsync(TestEntity2(id: 1), mode: PutMode.insert)
+            .timeout(defaultTimeout),
+        1);
+    expect(
+        () async => await box
+            .putAsync(TestEntity2(id: 5), mode: PutMode.update)
+            .timeout(defaultTimeout),
+        throwsA(predicate((ObjectBoxException e) =>
+            e.toString().contains('object with the given ID not found'))));
+    expect(box.count(), 1);
+
+    expect(
+        () async => await box
+            .putAsync(TestEntity2(id: 1), mode: PutMode.insert)
+            .timeout(defaultTimeout),
+        throwsA(predicate((ObjectBoxException e) =>
+            e.toString().contains('object with the given ID already exists'))));
+
+    // TODO test unique constraint violation and that an insert doesn't assign
+    //      object.id in that case.
+  });
+
+  test('.putAsync many', () async {
+    final items = List.generate(1000, (i) => TestEntityNonRel.filled(id: 0));
+    final futures = items.map(store.box<TestEntityNonRel>().putAsync).toList();
+    print('${futures.length} futures collected');
+    final ids = await Future.wait(futures);
+    print('${ids.length} futures finished');
+    for (int i = 0; i < items.length; i++) {
+      expect(items[i].id, ids[i]);
+    }
+  });
+
+  test('.putQueued', () {
+    final box = store.box<TestEntityNonRel>();
+    final items = List.generate(1000, (i) => TestEntityNonRel.filled(id: 0));
+    store.awaitAsyncSubmitted();
+    final ids = items.map(box.putQueued).toList();
+    for (int i = 0; i < items.length; i++) {
+      expect(items[i].id, ids[i]);
+    }
```

Comment:
The store is torn down after each test, so it does not matter if the async queue has not completed, right?

But maybe still wait and assert box count?

Reply:
Side note: I assumed it waits before to avoid any influence from other tests.